### PR TITLE
A mapping proposal

### PR DESCRIPTION
**Pull request by brentzundel**
The PR attempts to describe a mapping from vc+jwt to credential+ld+json according to the resolution made during the miami face to face meeting.
It is based on code written by Orie, but any mistakes are mine.

***
<a href="https://pr-preview.s3.amazonaws.com/brentzundel/vc-jwt/pull/60.html" title="Last updated on Mar 16, 2023, 5:15 PM UTC (058cf7e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-jwt/60/93385bc...brentzundel:058cf7e.html" title="Last updated on Mar 16, 2023, 5:15 PM UTC (058cf7e)">Diff</a>

### CODE REVIEW

File: index.html
Comment (awoie) [inline index.html:809]:
```suggestion
If <code>jti</code> is present, set the value of <code>id</code> as the
```

I believe this was a typo and you wanted to use `jti` for the `id` of the VC.

File: index.html
Comment (awoie) [inline index.html:810]: According to [RFC8141](https://www.rfc-editor.org/rfc/rfc8141.html), to be valid URN, the `urn:` has to be followed by a namespace (registered with [IANA](https://www.iana.org/assignments/urn-namespaces/urn-namespaces.xhtml)).

File: index.html
Comment (awoie) [inline index.html:830]: Same as above, we should make sure it becomes a valid URN.

File: index.html
Comment (awoie) [inline index.html:833]: Should we also add `@vocab`, to make sure that the added JWT claims don't cause issues with JSON-LD parsers?

**Review by awoie — CHANGES_REQUESTED** [review]
Direction looks good to me but I found a few issues that should be discussed in the PR.

File: index.html
Comment (OR13) [reply index.html:809]: indeed.

File: index.html
Comment (OR13) [reply index.html:833]: We don't need to add `@vocab` for v2... because it is in the v2 context.

File: index.html
Comment (awoie) [reply index.html:810]: I believe we can define our own namespace, yep.

Comment (brentzundel) [inline index.html:810]:
```suggestion
concatenation of "urn:vc:" and the value of <code>iat</code>.
```

File: index.html
Comment (brentzundel) [inline index.html:830]:
```suggestion
the concatenation of "urn:vc:" and the value of <code>sub</code>.
```

Comment (brentzundel) [reply index.html]: changed in [e7b8855](https://github.com/w3c/vc-jwt/pull/60/commits/e7b8855702f1cfd378cea2df3c7f3f0223f3f608)

File: index.html
Comment (brentzundel) [reply index.html]: changed in [e7b8855](https://github.com/w3c/vc-jwt/pull/60/commits/e7b8855702f1cfd378cea2df3c7f3f0223f3f608)

File: index.html
Comment (brentzundel) [reply index.html]: changed in [e7b8855](https://github.com/w3c/vc-jwt/pull/60/commits/e7b8855702f1cfd378cea2df3c7f3f0223f3f608)

**Review by OR13 — APPROVED** [review]

**Review by mprorock — APPROVED** [review]

**Review by awoie — APPROVED** [review]
looks good now

File: examples/vc-2.0/verifiable-credential.json
Comment (awoie) [inline examples/vc-2.0/verifiable-credential.json:4]:
```suggestion
  "issuer": "urn:vc:987654321",
```

**Review by awoie — APPROVED** [review]
lgtm after fixing the example with my suggestion

File: examples/vc-2.0/verifiable-credential.json
Comment (awoie) [inline examples/vc-2.0/verifiable-credential.json:6]:
```suggestion
    "id": "urn:vc:1234567890",
```

**Comment by awoie** [comment]
@brentzundel one question though. how does the mapping distinguish between JWT claims that are VC metadata and claims about the credential subject which get added to the `credentialSubject`? We should probably mention that in that case the mapping has this limitation -> no status method, no credential schema, no new VC metadata, ...

**Comment by brentzundel** [comment]
@awoie in general I agree that the mapping proposed here is minimal and could be improved on. I think it meets a minimal bar and would hope it could be merged as is, with additional improvements being made via other PRs and associated issues.
Would that be acceptable to you?

**Review by TallTed — APPROVED** [review]
Appears to be viable.

I think it's important to use full media types, i.e., this mapping is from `application/vc+jwt` to `application/credential+ld+json`

**Comment by awoie** [comment]
> @awoie in general I agree that the mapping proposed here is minimal and could be improved on. I think it meets a minimal bar and would hope it could be merged as is, with additional improvements being made via other PRs and associated issues. Would that be acceptable to you?

I think that is acceptable, yes.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-03-08-vcwg#section2-1) on 2023-03-08

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.1. A mapping proposal (pr vc-jwt#60)

_See github pull request [vc-jwt#60](https://github.com/w3c/vc-jwt/pull/60)._




**Orie Steele:** so, for VCJWT, we have 2 open PRs. PR 60 open by our wonderful chair Brent.  
… PR 60 - initially changes requested, seems currently no changes requested..  
… the PR uses everyone's favorite programming language, English, to describe a mapping, and address some of the concerns.  
… from compact native representations to the VC data model. Please review..  
… it's been open for 5 days with no objections. we'll merge it after a week.  

</details>

File: index.html
Comment (mirceanis) [inline index.html:833]: This implies that other top-level VC properties currently defined in the data model, such as `credentialSchema`, `credentialStatus`, `evidence`, etc. will map to sub-properties of the `credentialSubject`.

I think there should be an exception for these known properties of the data model. Am I missing something?

File: index.html
Comment (mirceanis) [inline index.html:794]: This implies that a JWT VC with a `type` property will be mapped to `credentialSubject.type` and that the VCDM representation of it will only ever have the value `["VerifiableCredential"]` for the `type`.

Is this the intent here, or am I missing something?

(same logic applies to `@context`)

**Review by mirceanis — CHANGES_REQUESTED** [review]
I spotted some typos and left some questions where I think there is ambiguity

File: index.html
Comment (mirceanis) [reply index.html:833]: Oh, I see this was asked already (https://github.com/w3c/vc-jwt/pull/60#issuecomment-1455758283)

File: index.html
Comment (OR13) [reply index.html:829]: yes, all the fields the core data model says are "URLs" need to be made "IRIs" if they are not.

File: index.html
Comment (OR13) [reply index.html:794]: That is correct, if the issue to of the JWT wanted to produce `credential+ld+json` or `vc+ld+json` there are instructions for how to do that in this document.

File: index.html
Comment (brentzundel) [reply index.html]: Thanks for the review, fixed in [ed1b378](https://github.com/w3c/vc-jwt/pull/60/commits/ed1b3784bbc59c0379bd542c44e9109a0b5a59cd)

File: index.html
Comment (brentzundel) [reply index.html]: Thanks for the review, fixed in [ed1b378](https://github.com/w3c/vc-jwt/pull/60/commits/ed1b3784bbc59c0379bd542c44e9109a0b5a59cd)

**Comment by OR13** [comment]
@brentzundel would you mind adding some of the text from the previous attempts to define the mapping?

```html
<section id="appendix">
  <h2>Appendix</h2>
  <section>
    <h3>Example VC Data Model Mapping</h3>
    <p>This is one possible unidirectional mapping between 2.0 VC-JWTs and the VC Data Model; other such mappings are possible.</p>
```

**Review by selfissued — CHANGES_REQUESTED** [review]
I agree with Orie's suggestion in https://github.com/w3c/vc-jwt/pull/60#issuecomment-1464678031 to apply some of the applicable language about the mapping being an example only one of such possible mappings.  After that, I'll approve the PR.

Thanks for doing this, Brent,

File: index.html
Comment (brentzundel) [inline index.html:776]:
```suggestion
The following describes a mapping from <code>vc+jwt</code> to
<code>credential+ld+json</code>. This is one possible unidirectional mapping
between 2.0 VC-JWTs and the VC Data Model; other such mappings are possible.
```

**Comment by brentzundel** [comment]
> I agree with Orie's suggestion in [#60 (comment)](https://github.com/w3c/vc-jwt/pull/60#issuecomment-1464678031) to apply some of the applicable language about the mapping being an example only one of such possible mappings. After that, I'll approve the PR.
> 
> Thanks for doing this, Brent,

Added in [7222c41](https://github.com/w3c/vc-jwt/pull/60/commits/7222c4140eab1c84c8ff38165445c5cb25a0efef)

**Comment by brentzundel** [comment]
> @brentzundel would you mind adding some of the text from the previous attempts to define the mapping?
> 
> ```
> <section id="appendix">
>   <h2>Appendix</h2>
>   <section>
>     <h3>Example VC Data Model Mapping</h3>
>     <p>This is one possible unidirectional mapping between 2.0 VC-JWTs and the VC Data Model; other such mappings are possible.</p>
> ```

Added in [7222c41](https://github.com/w3c/vc-jwt/pull/60/commits/7222c4140eab1c84c8ff38165445c5cb25a0efef)

**Comment by brentzundel** [comment]
@mirceanis I believe your questions have been addressed, if not could you let me know what concrete changes you'd like to see on this PR?

File: index.html
Comment (TallTed) [inline index.html:788]:
```suggestion
To construct the VCDM-compliant JSON:
```

File: index.html
Comment (TallTed) [inline index.html:791]:
```suggestion
Set the value of <code>@context</code> to
```

File: index.html
Comment (TallTed) [inline index.html:795]:
```suggestion
Set the value of <code>type</code> to <code>["VerifiableCredential"]</code>.
```

File: index.html
Comment (TallTed) [inline index.html:810]:
```suggestion
If <code>jti</code> is present, set the value of <code>id</code> to the
```

File: index.html
Comment (TallTed) [inline index.html:815]:
```suggestion
If <code>nbf</code> is present, set the value of <code>validFrom</code> to the
<code>dateTime</code> obtained by converting the value of <code>nbf</code> from
```

File: index.html
Comment (TallTed) [inline index.html:821]:
```suggestion
If <code>exp</code> is present, set the value of <code>validUntil</code> to the
<code>dateTime</code> obtained by converting the value of <code>exp</code> from
```

File: index.html
Comment (TallTed) [inline index.html:826]:
```suggestion
Set the value of <code>credentialSubject</code> to an object that contains
```

File: index.html
Comment (TallTed) [inline index.html:830]:
```suggestion
If <code>sub</code> is present, set the value of the <code>id</code> property to
```

File: index.html
Comment (TallTed) [inline index.html:776]:
```suggestion
The following describes a mapping from <code>application/vc+jwt</code> to
<code>application/credential+ld+json</code>. This is one possible unidirectional mapping
```

File: index.html
Comment (TallTed) [inline index.html:798]:
```suggestion
Set the value of the <code>issuer</code> property to one of the following:
```

File: index.html
Comment (TallTed) [inline index.html:804]:
```suggestion
If the value of <code>iss</code> is not a URL, use the concatenation of "<code>urn:vc:</code>"
```

File: index.html
Comment (TallTed) [inline index.html:811]:
```suggestion
concatenation of "<code>urn:vc:</code>" and the value of <code>jti</code>.
```

File: index.html
Comment (TallTed) [inline index.html:845]:
```suggestion
              title="application/vc+jwt mapped to application/credential+ld+json"
```

File: index.html
Comment (TallTed) [inline index.html:849]:
```suggestion
        <p>A VC encoded as application/vc+jwt:</p>
          <pre data-include="./examples/vc-2.0/verifiable-credential-jwt.json" data-include-format="json"></pre>
        <p>The application/credential+ld+json after applying the mapping:</p>
```

**Review by TallTed — CHANGES_REQUESTED** [review]

File: index.html
Comment (brentzundel) [inline index.html:849]:
```suggestion
        <p>The application/vc+ld+json after applying the mapping:</p>
```

**Review by selfissued — APPROVED** [review]
LGTM - Thanks Brent

**Comment by OR13** [comment]
@mirceanis @TallTed do you have any further change suggestions or comments?

**Review by TallTed — APPROVED** [review]
LGTM

**Comment by OR13** [comment]
Changes have been suggested, accepted there are no change requests, merging this pull request.

**Pull request merged**
